test/integration/istio_http_integration_test_with_envoy_jwt_filter.cc

```diff
@@ -0,0 +1,563 @@
+  )";
+  // From
+  // https://github.com/istio/istio/blob/master/security/tools/jwt/samples/jwks.json
+  constexpr char kJwksInline[] =
+      "{ \"keys\":[ "
+      "{\"e\":\"AQAB\",\"kid\":\"DHFbpoIUqrY8t2zpA2qXfCmr5VO5ZEr4RzHU_-envvQ\","
+      "\"kty\":\"RSA\",\"n\":\"xAE7eB6qugXyCAG3yhh7pkDkT65pHymX-"
+      "P7KfIupjf59vsdo91bSP9C8H07pSAGQO1MV"
+      "_xFj9VswgsCg4R6otmg5PV2He95lZdHtOcU5DXIg_"
+      "pbhLdKXbi66GlVeK6ABZOUW3WYtnNHD-91gVuoeJT_"
+      "DwtGGcp4ignkgXfkiEm4sw-4sfb4qdt5oLbyVpmW6x9cfa7vs2WTfURiCrBoUqgBo_-"
+      "4WTiULmmHSGZHOjzwa8WtrtOQGsAFjIbno85jp6MnGGGZPYZbDAa_b3y5u-"
+      "YpW7ypZrvD8BgtKVjgtQgZhLAGezMt0ua3DRrWnKqTZ0BJ_EyxOGuHJrLsn00fnMQ\"}]}";
+
+  return fmt::sprintf(kJwtFilterTemplate, Extensions::HttpFilters::HttpFilterNames::get().JwtAuthn,
+                      StringUtil::escape(kJwksInline),
+                      StringUtil::escape(kJwksInline));
+}
+
+std::string MakeAuthFilterConfig() {
+  constexpr char kAuthnFilterWithJwtTemplate[] = R"(
+    name: %s
+    config:
+      policy:
+        origins:
+        - jwt:
+            issuer: testing@secure.istio.io
+            jwks_uri: http://localhost:8081/
+        - jwt:
+            issuer: testing-rbac@secure.istio.io
+            jwks_uri: http://localhost:8081/
+        principalBinding: USE_ORIGIN
+)";
+  return fmt::sprintf(kAuthnFilterWithJwtTemplate,
+                      Utils::IstioFilterName::kAuthentication);
+}
+
+std::string MakeRbacFilterConfig() {
+  constexpr char kRbacFilterTemplate[] = R"(
+  name: envoy.filters.http.rbac
+  config:
+    rules:
+      policies:
+        "foo":
+          permissions:
+            - any: true
+          principals:
+            - metadata:
+                filter: %s
+                path:
+                  - key: %s
+                value:
+                  string_match:
+                    exact: %s
+)";
+  return fmt::sprintf(
+      kRbacFilterTemplate, Utils::IstioFilterName::kAuthentication,
+      istio::utils::AttributeName::kRequestAuthPrincipal, kExpectedPrincipal);
+}
+
+std::string MakeMixerFilterConfig() {
+  constexpr char kMixerFilterTemplate[] = R"(
+  name: mixer
+  config:
+    defaultDestinationService: "default"
+    mixerAttributes:
+      attributes: {
+      }
+    serviceConfigs: {
+      "default": {}
+    }
+    transport:
+      attributes_for_mixer_proxy:
+        attributes: {
+          "source.uid": {
+            string_value: %s
+          }
+        }
+      report_cluster: %s
+      check_cluster: %s
+  )";
+  return fmt::sprintf(kMixerFilterTemplate, kSourceUID, kTelemetryBackend,
+                      kPolicyBackend);
+}
+
+
+// checkJwtRawClaim finds the serialized jwt payload and check to see if all key-value pairs
+// from the jwt claim is there.
+// Returns false if it cannot find the serialized jwt payload or if any key-value pair does not
+// matched the expectation.
+bool checkJwtRawClaim(const ::google::protobuf::RepeatedPtrField<::std::string>& words) {
```

Comment:
It's ok for now, but may be good to know why upstream JWT filter change the order of claims in the raw payload.

+ @qiwzhang Do you have any insight?

Reply:
I emailed Wayne this morning but he's OOO until next week.